### ISSUE TITLE
Kotlin 2.1 Compatibility Issue

### ISSUE THREAD

**Issue by hantsy** [pinterest/ktlint#2882]
After several hours of research yesterday, I found the Kotlinter gradle broke our project build when upgrading Kotlin 2.1. 

I tried to fork and update Kotin Kotlinter to Kotlin 2.1 but it did not resolve the issue. 

The root cause is from Klint, check the issue discussion here: https://github.com/jeremymailen/kotlinter-gradle/issues/414#issuecomment-2505662213

```bash
> Task :lintKotlinIntegration FAILED
lint worker execution error
java.lang.NoClassDefFoundError: org/jetbrains/kotlin/gradle/internal/config/CompilerConfigurationKey
	at org.jetbrains.kotlin.compiler.plugin.ComponentRegistrar$Companion.<clinit>(ComponentRegistrar.kt:33)
	at org.jetbrains.kotlin.compiler.plugin.ComponentRegistrar.<clinit>(ComponentRegistrar.kt)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment$Companion.registerExtensionsFromPlugins$cli_base(KotlinCoreEnvironment.kt:753)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment$ProjectEnvironment.registerExtensionsFromPlugins(KotlinCoreEnvironment.kt:194)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment$Companion.configureProjectEnvironment(KotlinCoreEnvironment.kt:652)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment.<init>(KotlinCoreEnvironment.kt:224)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment.<init>(KotlinCoreEnvironment.kt)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment$Companion.createForProduction(KotlinCoreEnvironment.kt:468)
	at com.pinterest.ktlint.rule.engine.internal.KotlinPsiFileFactoryKt.initPsiFileFactory(KotlinPsiFileFactory.kt:58)
	at com.pinterest.ktlint.rule.engine.internal.KotlinPsiFileFactoryProvider.getKotlinPsiFileFactory(KotlinPsiFileFactory.kt:32)
	at com.pinterest.ktlint.rule.engine.internal.RuleExecutionContext$Companion.createRuleExecutionContext$ktlint_rule_engine(RuleExecutionContext.kt:212)
	at com.pinterest.ktlint.rule.engine.internal.CodeFormatter.format(CodeFormatter.kt:51)
	at com.pinterest.ktlint.rule.engine.internal.CodeFormatter.format(CodeFormatter.kt:28)
	at com.pinterest.ktlint.rule.engine.api.KtLintRuleEngine.lint(KtLintRuleEngine.kt:91)
	at org.jmailen.gradle.kotlinter.tasks.lint.LintWorkerAction.execute(LintWorkerAction.kt:49)
```

**Comment by paul-dingemans** [comment]
Ktlint uses the Kotlin 2.0.21 embedded compiler. https://github.com/pinterest/ktlint/pull/2880 also shows that upgrade to 2.1 is failing as some keywords have been removed which did exist in 2.0.21.

**Comment by hantsy** [comment]
Besides calling the legacy method, you should also consider handling the dependency `kotlin-compiler-embeddable`. When upgrading to Kotlin 2.1.0, a warning is raised when building the project.

Check: https://kotlinlang.org/docs/whatsnew21.html#compiler-symbols-hidden-from-the-kotlin-gradle-plugin-api

**Comment by paul-dingemans** [comment]
> Besides calling the legacy method, you should also consider handling the dependency `kotlin-compiler-embeddable`. When upgrading to Kotlin 2.1.0, a warning is raised when building the project.
> 
> Check: https://kotlinlang.org/docs/whatsnew21.html#compiler-symbols-hidden-from-the-kotlin-gradle-plugin-api

Which legacy method do you mean?

**Comment by hantsy** [comment]
> Which legacy method do you mean?

The exceptions caused the build failure.

**Comment by paul-dingemans** [comment]
Please keep in mind that I don't run kotlinter (nor any other Ktlint integrator). So I cannot verify myself whether the problem is solved if I do not know exactly what to resolve. So either you have to provide me with a reproduction path with Ktlint CLI, or with enough detailed information so that I can check with enough certainity that it is resolved.

**Comment by paul-dingemans** [comment]
For now, I am assuming that this has been validated in the [kolinter issue](https://github.com/jeremymailen/kotlinter-gradle/issues/414#issuecomment-2510977000) which was fixed by PR #2880
